src/iamserver/nodeinfoprovider.cpp

```diff
@@ -0,0 +1,313 @@
+            PopulateCPUInfoObject();
+        } catch (...) {
+            LOG_ERR() << "Failed to parse CPU info file: line = " << line.c_str();
+
+            return aos::ErrorEnum::eFailed;
+        }
+
+        return aos::ErrorEnum::eNone;
+    }
+
+    std::ifstream                            mFile;
+    std::unordered_map<size_t, aos::CPUInfo> mCPUInfos;
+    std::vector<KeyValue>                    mCurrentEntryKeyValues;
+};
+
+aos::RetWithError<uint64_t> GetMemoryInfo(const std::string& path)
+{
+    const std::string memInfoPath = path.empty() ? cMemInfoPath : path;
+
+    std::ifstream file;
+    if (file.open(memInfoPath); !file.is_open()) {
+        return {0, aos::ErrorEnum::eNotFound};
+    }
+
+    std::string line;
+    while (std::getline(file, line)) {
+        const auto keyValue = ParseKeyValue(line);
+
+        if (!keyValue.has_value() || keyValue->mKey != "MemTotal") {
+            continue;
+        }
+
+        const auto memTotalKB = std::stoull(keyValue->mValue.substr(0, keyValue->mValue.find(" ")));
+
+        // convert KB to bytes
+        return {memTotalKB * cBytesPerKB, aos::ErrorEnum::eNone};
+    }
+
+    return {0, aos::ErrorEnum::eNotFound};
+}
+
+uint64_t GetTotalSize(std::string_view path)
+{
+    struct statfs stat { };
+
+    if (statfs(path.data(), &stat) == -1) {
+        return 0;
+    }
+
+    return stat.f_blocks * stat.f_bsize;
+}
+
+aos::RetWithError<aos::NodeStatus> GetNodeStatus(const std::string& path)
+{
+    const std::string provisioningStatusPath = path.empty() ? cProvisioningStatusPath : path;
+
+    std::ifstream file;
+    if (file.open(provisioningStatusPath); !file.is_open()) {
+        return {aos::NodeStatusEnum::eUnprovisioned, aos::ErrorEnum::eNotFound};
+    }
+
+    std::string line;
+    std::getline(file, line);
+
+    aos::NodeStatus nodeStatus;
+    auto            err = nodeStatus.FromString(line.c_str());
+
+    return {nodeStatus, err};
+}
+
+} // namespace
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+aos::Error NodeInfoProvider::Init(const NodeInfoConfig& config)
+{
+    mCPUInfoPath = config.mCPUInfoPath.empty() ? cCPUInfoPath : config.mCPUInfoPath;
+    mMemInfoPath = config.mMemInfoPath.empty() ? cMemInfoPath : config.mMemInfoPath;
+    mProvisioningStatusPath
+        = config.mProvisioningStatePath.empty() ? cProvisioningStatusPath : config.mProvisioningStatePath;
+
+    mNodeInfo.mID     = config.mNodeID.c_str();
+    mNodeInfo.mType   = config.mNodeType.c_str();
+    mNodeInfo.mName   = config.mNodeName.c_str();
+    mNodeInfo.mOSType = config.mOSType.c_str();
+
+    if (const auto [totalRam, err] = GetMemoryInfo(config.mMemInfoPath); err.IsNone()) {
+        mNodeInfo.mTotalRAM = totalRam;
+    } else {
+        return AOS_ERROR_WRAP(err);
+    }
```

Comment:
in my opinion will be better to handle errors and return without else, the reason why I think more readable and it more convenient for our style. But it's up to you)